core/src/main/scala/org/apache/spark/CacheRecoveryManager.scala

```diff
@@ -0,0 +1,250 @@
+  private val scheduler =
+    ThreadUtils.newDaemonSingleThreadScheduledExecutor("recover-cache-shutdown-timers")
+
+    private val blocksToSave = new mutable.HashMap[String, mutable.PriorityQueue[RDDBlockId]]
+    private val savedBlocks = new mutable.HashMap[String, mutable.HashSet[RDDBlockId]]
+    private val killTimers = new mutable.HashMap[String, ScheduledFuture[_]]
+
+  /**
+   * Query Block Manager Master for cached blocks.
+   *
+   * @param execIds the executors to query
+   * @return a map of executorId to its replication state.
+   */
+  def getBlocks(execIds: Seq[String]): Map[String, ExecutorReplicationState] = synchronized {
+    val isThereEnoughMemory = checkFreeMem(execIds)
+
+    isThereEnoughMemory.map {
+      case (execId, true) =>
+        updateBlockState(execId)
+        val blocksRemain = blocksToSave.get(execId).exists(_.nonEmpty)
+        execId -> (if (blocksRemain) HasCachedBlocks else NoMoreBlocks)
+      case (execId, false) =>
+        execId -> NotEnoughMemory
+    }
+  }
+
+  /**
+   * Checks to see if there is enough memory on the cluster to hold the cached blocks on these
+   * executors. Checks on a per executor basis from the smallest to the largest.
+   *
+   * @param execIds the executors to query
+   * @return a map of executor ids to whether or not there is enough free memory for them.
+   */
+  private def checkFreeMem(execIds: Seq[String]): Map[String, Boolean] = {
+    val execIdsToShutDown = execIds.toSet
+    val allExecMemStatus: Map[String, (Long, Long)] = blockManagerMasterEndpoint
+      .askSync[Map[BlockManagerId, (Long, Long)]](GetMemoryStatus)
+      .map { case (blockManagerId, mem) => blockManagerId.executorId -> mem }
+
+    val (expiringMemStatus, remainingMemStatus) = allExecMemStatus.partition {
+      case (k, v) => execIdsToShutDown.contains(k)
+    }
+    val freeMemOnRemaining = remainingMemStatus.values.map(_._2).sum
+
+    val alreadyReplicatedBlocks: Map[String, Set[RDDBlockId]] =
+      savedBlocks.filterKeys(execIdsToShutDown).map { case (k, v) => k -> v.toSet }.toMap
+
+    val getSizes = GetSizeOfBlocks(alreadyReplicatedBlocks)
+    val alreadyReplicatedBytes = blockManagerMasterEndpoint.askSync[Map[String, Long]](getSizes)
+
+    val bytesToReplicate: Seq[(String, Long)] =
+      expiringMemStatus.map { case (execId, (maxMem, remainingMem)) =>
+        val alreadyReplicated = alreadyReplicatedBytes.getOrElse(execId, 0L)
+        val usedMem = maxMem - remainingMem
+        val toBeReplicatedMem = usedMem - alreadyReplicated
+        execId -> toBeReplicatedMem
+      }.toSeq.sortBy { case (k, v) => v }
+
+    bytesToReplicate.scan(("start", freeMemOnRemaining)) {
+      case ((_, remainingMem), (execId, mem)) => (execId, remainingMem - mem)
+    }.drop(1)
+      .toMap
+      .mapValues(freeMem => freeMem >= 0)
+  }
+
+  /**
+   * Gets a list of cached blocks from the block manager master and updates.
```

Comment:
"Updates the internal state to track the blocks from the given executor that need recovery."

Always start with what the method's purpose, not how it achieves that.

Reply:
removed this method